### PR TITLE
introduce linkat_replace() helper, and port various things over to it

### PR DESCRIPTION
**Pull request by poettering**
This is preparation for #26826, but makes a ton of sense on its own.

### CODE REVIEW

File: src/basic/fs-util.c [hunk 1]
```diff
@@ -1256,3 +1256,101 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
...
+        int r;
+
+        /* Like linkat() but replaces the target if needed. Is a NOP if source and target already share the
+         * same inode. */
+
+        if (olddirfd == AT_FDCWD && isempty(oldpath)) /* Refuse operating on the cwd (which is a dir, and dirs can't be hardlinked) */
+                return -EISDIR;
+
+        if (oldpath) {
+                if (STR_IN_SET(oldpath, ".", "..")) /* Refuse these definite directories early */
```
Comment (YHNdnzj) [inline src/basic/fs-util.c]: `dot_or_dot_dot`?

File: src/basic/fs-util.c [hunk 2]
```diff
@@ -1256,3 +1256,101 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
 
         return RET_NERRNO(linkat(fd, "", newdirfd, newpath, AT_EMPTY_PATH));
 }
+
+int linkat_replace(int olddirfd, const char *oldpath, int newdirfd, const char *newpath) {
+        _cleanup_close_ int old_fd = -EBADF;
+        int r;
+
```
Comment (YHNdnzj) [inline src/basic/fs-util.c:1263]:
```c
assert(olddirfd >= 0 || olddirfd == AT_FDCWD);
assert(newdirfd >= 0 || newdirfd == AT_FDCWD);
```

File: src/basic/fs-util.c [hunk 3]
```diff
@@ -1256,3 +1256,101 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
...
+        }
+
+        if (isempty(newpath)) /* source path is optional, but the target path is not */
+                return -EINVAL;
+
+        if (STR_IN_SET(newpath, ".", ".."))
+                return -EISDIR;
+
+        if (ENDSWITH_SET(newpath, "/", "/.", "/..")) /* ditto */
+                return -EISDIR;
```
Comment (YHNdnzj) [inline src/basic/fs-util.c]: It would be nicer to extract this into a helper in `path-util`?

File: src/basic/fs-util.c [hunk 4]
```diff
@@ -1256,3 +1256,101 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
...
+
+                /* Second, let's try to go via the opened fd as source with AT_EMPTY_PATH. (This requires
+                 * CAP_DAC_READ_SEARCH and a more recent kernel, but does not require /proc/ mounted.) */
+                r = RET_NERRNO(linkat(old_fd, "", newdirfd, tmp_path, AT_EMPTY_PATH));
+                if (r < 0) {
+                        /* linkat() with AT_EMPTY_PATH requires privs. Let's cover for that. Interestingly,
+                         * if we lack the privs the kernel returns ENOENT rather than EACCESS/EPERM. But
+                         * let's also check for the error here that actually makese sense */
+                        if (!ERRNO_IS_PRIVILEGE(r) && r != -ENOENT)
+                                return r;
```
Comment (YHNdnzj) [inline src/basic/fs-util.c]: Hmm, this is `link_fd`?

File: src/basic/fs-util.c [hunk 5]
```diff
@@ -1256,3 +1256,101 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
...
+        if (r < 0)
+                return r;
+
+        /* Try hardlink via /proc/self/fd/ first, which requires the least privileges, but does require /proc/ to be mounted. */
+        r = RET_NERRNO(linkat(AT_FDCWD, FORMAT_PROC_FD_PATH(old_fd), newdirfd, tmp_path, AT_SYMLINK_FOLLOW));
+        if (r < 0) {
+                if (r == -ENOENT) {
+                        if (proc_mounted() != 0) /* Unless ENOENT is caused by /proc/ is not being around, let's not bother with a fallback */
+                                return r;
+                } else if (!ERRNO_IS_PRIVILEGE(r))
```
Comment (YHNdnzj) [inline src/basic/fs-util.c]: EPERM/EACCES aren't checked in `link_fd` but checked here? And particular reason?

File: src/basic/fs-util.c [hunk 6]
```diff
@@ -1256,3 +1256,101 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
...
+
+        old_fd = xopenat(olddirfd, oldpath, O_PATH|O_CLOEXEC);
+        if (old_fd < 0)
+                return old_fd;
+
+        struct stat old_st;
+        if (fstat(old_fd, &old_st) < 0)
+                return -errno;
+
+        if (S_ISDIR(old_st.st_mode)) /* Don't bother if are operating on a directory */
```
Comment (yuwata) [inline src/basic/fs-util.c]: `if are` -> `if we are` ?

File: src/basic/fs-util.c (hunk 3 above)
Comment (poettering) [reply src/basic/fs-util.c]: indeed

File: src/basic/fs-util.c (hunk 4 above)
Comment (poettering) [reply src/basic/fs-util.c]: indeed

**Review by bluca — APPROVED** [review]

File: src/basic/fs-util.c [hunk 7]
```diff
@@ -1256,3 +1256,81 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
...
+        assert(olddirfd >= 0 || olddirfd == AT_FDCWD);
+        assert(newdirfd >= 0 || newdirfd == AT_FDCWD);
+
+        /* Like linkat() but replaces the target if needed. Is a NOP if source and target already share the
+         * same inode. */
+
+        if (olddirfd == AT_FDCWD && isempty(oldpath)) /* Refuse operating on the cwd (which is a dir, and dirs can't be hardlinked) */
+                return -EISDIR;
+
+        if (oldpath && path_implies_directory(oldpath)) /* Refuse these definite directories early */
```
Comment (YHNdnzj) [inline src/basic/fs-util.c]: `path_implies_directory` returns false if path == NULL, so the `oldpath` condition is not needed?

File: src/basic/fs-util.c [hunk 8]
```diff
@@ -1256,3 +1256,81 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
...
+                return -EISDIR;
+
+        if (isempty(newpath)) /* source path is optional, but the target path is not */
```
Comment (YHNdnzj) [inline src/basic/fs-util.c]: This should be an assertion?

File: src/basic/fs-util.c [hunk 9]
```diff
@@ -1256,3 +1256,81 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
...
+        _cleanup_close_ int old_fd = -EBADF;
+        int r;
+
+        assert(olddirfd >= 0 || olddirfd == AT_FDCWD);
+        assert(newdirfd >= 0 || newdirfd == AT_FDCWD);
+
+        /* Like linkat() but replaces the target if needed. Is a NOP if source and target already share the
+         * same inode. */
+
+        if (olddirfd == AT_FDCWD && isempty(oldpath)) /* Refuse operating on the cwd (which is a dir, and dirs can't be hardlinked) */
```
Comment (YHNdnzj) [inline src/basic/fs-util.c:1271]: (I think this can be an assertion too. But doesn't matter)

File: src/basic/fs-util.c [hunk 10]
```diff
@@ -1256,3 +1256,81 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
...
+                return -errno;
+
+        if (S_ISDIR(new_st.st_mode)) /* Refuse replacing directories */
+                return -EEXIST;
+
+        if (stat_inode_same(&old_st, &new_st)) /* Already the same inode? Then shortcut this */
+                return 0;
+
+        _cleanup_free_ char *tmp_path = NULL;
+        r = tempfn_random(newpath, /* extra */ NULL, &tmp_path);
```
Comment (YHNdnzj) [inline src/basic/fs-util.c]: `extra =`

**Review by YHNdnzj — APPROVED** [review]

**Review by yuwata — APPROVED** [review]

**Pull request merged**